### DISCUSSION TITLE
Complie CoolProp in Visual Studio 2013.

### DISCUSSION THREAD

**Discussion by lihy213** [CoolProp/CoolProp#2304] [Q&A]
Recently, I tried to embed CoolProp into Visual Studio 2013 for C++ development, but there seems no guidance online and the CoolProp repository has no lib document. Does there anybody know how to solve it? 
Kindly regards.

**Comment by ibell** [comment]
Yes, you can see the instructions here for building a static library: http://www.coolprop.org/coolprop/wrappers/StaticLibrary/index.html#static-library

And a complete example in this repo for linking with your own code: https://github.com/CoolProp/simple_cmake_example

**Reply by lihy213** [reply]
Thanks for the link. I followed the link step by step, but there's also something wrong "can't open files: 'fmt/format.h', No such file or directory", are there any other solutions?

**Reply by henningjp** [reply]
When you clone the repository, use --recursive to get all of the external modules (you are probably doing this).  Then make sure that in the externals/fmtlib directory that the required commit is checked out for the version/release of CoolProp that you have checked out.  The current (as of release 6.5.0) fmtlib commit needed is:

![image](https://github.com/CoolProp/CoolProp/assets/17114032/6ffc8251-785a-49c8-bc7c-5b314aaa26f8)

Earlier releases of CoolProp used commit [135ab5c](https://github.com/fmtlib/fmt/tree/135ab5cf71ed731fc9fa0653051e7d4884a3652f) of fmtlib.  Look for the checkout commit required on Github or using the git modules command and then checkout that commit in the fmtlib directory.

This was a recent change and one I stumble over all the time.

**Comment by henningjp** [comment]
@lihy213  Alternatively, take a look at the [Mathcad wrapper](https://github.com/CoolProp/CoolProp/tree/master/wrappers/MathCAD).  There is a single C++ file that builds a DLL and links CoolProp in statically - no external CoolProp DLL required.

In the initialization section of the code, there are three includes:
```cpp
#include "CoolProp.h"
#include "DataStructures.h"
#include "HumidAirProp.h"
```

Then the code makes direct calls to CoolProp functions in the CoolProp namespace:
```cpp
[...]

    // pass the arguments to the CoolProp.Props1() function
    Prop->real = CoolProp::Props1SI(Fluid->str, PropName->str);

    // Note: Props1SI does not throw exceptions, but instead
    // sets global parameter "errstring" and returns _HUGE.
    // Use ValidNumber(val) to see if Props1SI failed with an error message...
    if (!ValidNumber(Prop->real)) {
        std::string emsg = CoolProp::get_global_param_string("errstring");
        CoolProp::set_error_string(emsg);  // reset error string so Mathcad can retrieve it

[...]
```
This might work for you if you don't mind the whole of CoolProp being statically linked within your .exe or .dll.
